website: Add 3.0 blog post (#4046)
transloadit/uppy: website/src/_posts/2022-09-3.0.md

---
title: "Uppy 3.0: Future proof, conveniently easy, loyally stable" 
date: 2022-09-01
author: 
  - aduh95
  - arturi
  - mifi
  - murderlon
image: "https://uppy.io/images/blog/3.0/uppy-3-0.jpg"
published: true
---

![](/images/blog/3.0/uppy-3-0.jpg)

Uppy is turning 3 (or 29 in [dog years](https://www.akc.org/expert-advice/health/how-to-calculate-dog-years-to-human-years/)). Proper adulthood, that is, as it’s ready to stay loyal but without the silly mistakes (read: bugs), with an (ESM) surgery to keep strolling by your side in the current ecosystem, and other behaviour improvements 🐶

<!--more-->

## ESM surgery

> [ECMAScript Modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) (ESM) are the future and we’re ready to adopt it. It’s benefits over CommonJS include: being more secure, better at tree shaking, and its syntax can be understood natively by browsers, which can greatly improve the DX if the tools are configured to take advantage of that.

Reviewer: ```suggestion
[ECMAScript Modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) (ESM) are the future and we’re ready to adopt it. Its benefits over CommonJS include: being more secure, better at tree shaking, and its syntax can be understood natively by browsers, which can greatly improve the DX if the tools are configured to take advantage of that.
```

Following the footsteps of many packages, we now only ship Uppy core and its plugins as ESM. On Uppy 2.x, we were shipping CommonJS.

If are already using ESM yourself, or are using the CDN builds, nothing changes for you! If you are using CommonJS, you might need to add some tooling for everything to work, or you might want to refactor your codebase to ESM – refer to the [Pure ESM package gist](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) for added information and help on how to do that.

## Less integration code with the new remote sources preset

> We’ve introduced a new Uppy preset plugin, [`@uppy/remote-sources`](/docs/remote-sources), which combines Instagram, Facebook, Google Drive, Box, Unsplash, Dropbox, One Drive, Zoom and any future remote cloud sources that Uppy will support. It only works with Dashboard and let’s you enable all the above sources in only line.

Reviewer: ```suggestion
We’ve introduced a new Uppy preset plugin, [`@uppy/remote-sources`](/docs/remote-sources), which combines Instagram, Facebook, Google Drive, Box, Unsplash, Dropbox, One Drive, Zoom and any future remote cloud sources that Uppy will support. It only works with Dashboard and lets you enable all the above sources in only line.
```

Example code:

```js
const { app, emitter } = companion.app(options)

emitter.on('upload-start', ({ token }) => {
  console.log('Upload started', token)

  function onUploadEvent ({ action, payload }) {
    if (action === 'success') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.log('Upload finished', token, payload.url)
    } else if (action === 'error') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.error('Upload failed', payload)
    }
  }
  emitter.on(token, onUploadEvent)
})
```

### Internal improvements

Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. This will lead to more less bugs and security issues. It will also make it easier to implement new custom Providers.

We have also replaced the deprecated [request](https://github.com/request/request) library with [got](https://github.com/sindresorhus/got). This also removed the need for [purest](https://github.com/simov/purest).

### And more!

> * [Compressor](/docs/compressor) is officially stable. The Compressor plugin for Uppy optimizes images (JPEG, PNG), saving on average up to 60% in size (roughly 18 MB for 10 images). It’s been battle tested and has proved itself.

Reviewer: > roughly 18 MB for 10 images

Images are very different in terms of file size, so I'd propose to make the estimate more specific (e.g: 18 MB of random 10 iPhone JPG photos), or just remove the estimate at all — "Up to 60% in size" is already easy to understand.

---
title: "Uppy 3.0: Future proofed, fresh, loyally familiar" 
date: 2022-09-01
author: 
  - aduh95
  - arturi
  - mifi
  - murderlon
image: ""
published: true
---

<!--more-->

## ESM

Following the footsteps of many packages, we now only ship Uppy core and its plugins as ECMAScript Modules (ESM). On Uppy 2.x, we were shipping CommonJS.

If are already using ESM yourself, or are using the CDN builds, nothing changes for you! If you are using CommonJS, you might need to add some tooling for everything to work, or you might want to refactor your codebase to ESM – refer to the [Pure ESM package gist](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) for added information and help on how to do that.

> <img class="border" alt="Uppy Compressor plugin showing compressed images notification" src="/images/blog/2.4-2.7/compressor.jpg">

Reviewer 2: Compressor image under ESM?

## Remote Sources

We’ve introduced a new Uppy plugin/preset that combines Instagram, Facebook, Google Drive, Box, Unsplash, Dropbox, One Drive, Zoom and any future remote cloud sources that Uppy will support. It only works with Dashboard and let’s you enable all the above sources in only line:

```js
uppy.use(RemoteSoruces, { companionUrl: 'https://companion.uppy.io' }) // Remote Soruces need Companion server instance to work, you can get one hosted from Transloadit.com or self-host it

> ```

Reviewer 2: I think it would be nice to show a before and after code snippet to further highlight the improvement

But, it didn’t come without its own set of new problems:

* It tries to do the exact same, but it looks like a different product.
* It’s confusing for users whether they want to use Robodog or Uppy directly.
* Robodog is more ergonomic because it’s limited. When you hit such a limit, you
  need to refactor everything to Uppy with plugins.

This has now led us to deprecating Robodog and embrace Uppy for its strong suits: modularity and flexibility. At the same time, we also introduced something to take away some repetitive integration code: [`@uppy/remote-sources`](/docs/remote-sources).

To mimic the Robodog implementation with all its features, you can use the code snippet below. But chances are Robodog did more than you need so feel free to remove things or go through the [list of plugins](/docs/plugins/) and install and use the ones you need.

You can also checkout how we migrated the Robodog example ourselves in this [commit](https://github.com/transloadit/uppy/commit/089aaed615c77bafaf905e291b6b4e82aaeb2f6f).

```js

> // Import just the plugins you need
> import Uppy from '@uppy/core'
> import Dashboard from '@uppy/dashboard'
> import RemoteSources from '@uppy/remote-sources'
> import Webcam from '@uppy/webcam'
> import ScreenCapture from '@uppy/screen-capture'
> import ImageEditor from '@uppy/image-editor'
> import Audio from '@uppy/audio'
> import Transloadit, { COMPANION_URL, COMPANION_ALLOWED_HOSTS } from '@uppy/transloadit'
> 
> // Import their styles
> import '@uppy/core/dist/style.css'
> import '@uppy/dashboard/dist/style.css'
> import '@uppy/webcam/dist/style.css'
> import '@uppy/audio/dist/style.css'
> import '@uppy/screen-capture/dist/style.css'
> import '@uppy/image-editor/dist/style.css'
> 
> // Initialize Uppy

Reviewer 2: nit: these code comments are too obvious to add any value. It's like have a comment `// variable` above `const foo = 'bar'`.

```suggestion
import Uppy from '@uppy/core'
import Dashboard from '@uppy/dashboard'
import RemoteSources from '@uppy/remote-sources'
import Webcam from '@uppy/webcam'
import ScreenCapture from '@uppy/screen-capture'
import ImageEditor from '@uppy/image-editor'
import Audio from '@uppy/audio'
import Transloadit, { COMPANION_URL, COMPANION_ALLOWED_HOSTS } from '@uppy/transloadit'

import '@uppy/core/dist/style.css'
import '@uppy/dashboard/dist/style.css'
import '@uppy/webcam/dist/style.css'
import '@uppy/audio/dist/style.css'
import '@uppy/screen-capture/dist/style.css'
import '@uppy/image-editor/dist/style.css'

```

const uppy = new Uppy()
  .use(Dashboard, { inline: true, target: '#app' })
  .use(RemoteSources, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(Webcam, { target: Dashboard })
  .use(Audio, { target: Dashboard })
  .use(ScreenCapture, { target: Dashboard })
  .use(ImageEditor, { target: Dashboard })
  .use(Transloadit, {
    async getAssemblyOptions (file) {
      // This is where you configure your auth key, auth secret, and template ID
      // https://uppy.io/docs/transloadit/#getAssemblyOptions-file
      //
      // It is important to set the secret in production:
      // https://transloadit.com/docs/topics/signature-authentication/
      const response = await fetch('/some-endpoint')
      return response.json()
    },
  })
```

## Companion 4.0

### Streaming upload

> Streaming upload can now be enabled in Companion with Tus. This gives greatly improved upload speeds when using remote Providers and allows uploading potentially hundreds of gigabytes without needing a large server storage.

Reviewer 2: I think it would be nice to add one sentence on why it only works with Tus?

Reviewer 3: Ah, I will make it more clear that it now **also** works with Tus. In the previous version it worked with all destinations except Tus.

```suggestion
Streaming upload can now be enabled in Companion also when using Tus. This gives greatly improved upload speeds when using remote Providers and allows uploading potentially hundreds of gigabytes without needing a large server storage. Note that this feature was also available before Companion 4, but did not work with Tus.
```

With streaming upload disabled (default), the whole file will be downloaded first, and only when the download has completely finished would the upload start.

```js
// Pseudo-code:
const providerStream = provider.getFileStream()
const tempFilePath = 'tmpfile'
const fileStream = fs.createWriteStream(tempFilePath)
await pipeline(providerStream, fileStream)
await tus.uploadFile(tempFilePath)
```

When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.

```js
// Pseudo-code:
const providerStream = provider.getFileStream()
await tus.uploadStream(providerStream)

> ```

Reviewer 2: These code snippets show internals which AFAIK are not valuable or necessarily interesting for a blog post reader.

Reviewer 3: I thought so too, but @arturi suggested that more details are better, because the readers are devs, see https://github.com/transloadit/uppy/issues/4042#issuecomment-1228423538

Reviewer 2: I understand, but more details don't mean any details. It's pseudo internal code, I just don't think it adds anything here. If we want more details, the way to go is adding stats about the speed improvements.

> 
> ```js
> // Pseudo-code:
> const providerStream = provider.getFileStream()
> const tempFilePath = 'tmpfile'
> const fileStream = fs.createWriteStream(tempFilePath)
> await pipeline(providerStream, fileStream)
> await tus.uploadFile(tempFilePath)
> ```
> 
> When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.
> 
> ```js
> // Pseudo-code:
> const providerStream = provider.getFileStream()
> await tus.uploadStream(providerStream)
> ```

Reviewer 3: ```suggestion

When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.
```

For more information, [see the docs](https://uppy.io/docs/companion/).

### Event emitter

Companion now has a server-side event emitter that can be used to detect when uploads start, finish and fail, without having to depend on the client (so it works even if the user has closed their browser).

Example code:

```js
const { app, emitter } = companion.app(options)

emitter.on('upload-start', ({ token }) => {
  console.log('Upload started', token)

  function onUploadEvent ({ action, payload }) {
    if (action === 'success') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.log('Upload finished', token, payload.url)
    } else if (action === 'error') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.error('Upload failed', payload)
    }
  }
  emitter.on(token, onUploadEvent)
})
```

> ### `Async`/`await`-ify Companion

Reviewer 2: I think we should combine this with refactor from `request` to `got`, making this more generic (something like "internal improvements").

> ### `Async`/`await`-ify Companion

Reviewer 3: ```suggestion
### Internal improvements
```

Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. As can be seen [in this PR](https://github.com/transloadit/uppy/pull/3953), we lost 425 lines!

> For example the new facebook Provider implementation is only [100 lines of code](https://github.com/transloadit/uppy/blob/35812ca378b9b02290a2ab51261ddaeff8211fda/packages/%40uppy/companion/src/server/provider/facebook/index.js), [down from 194](https://github.com/transloadit/uppy/blob/7b0db7d3a0163374cfbde01e826c86ead4f0aacb/packages/%40uppy/companion/src/server/provider/facebook/index.js). This should make it much easier to implement new Providers.

Reviewer 2: I think this dives a bit too much into the internals. I think we should focus more on writing about how an internal change has X positive effect for the user of Companion. For instance, security improvement.

> Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. As can be seen [in this PR](https://github.com/transloadit/uppy/pull/3953), we lost 425 lines!
> 
> For example the new facebook Provider implementation is only [100 lines of code](https://github.com/transloadit/uppy/blob/35812ca378b9b02290a2ab51261ddaeff8211fda/packages/%40uppy/companion/src/server/provider/facebook/index.js), [down from 194](https://github.com/transloadit/uppy/blob/7b0db7d3a0163374cfbde01e826c86ead4f0aacb/packages/%40uppy/companion/src/server/provider/facebook/index.js). This should make it much easier to implement new Providers.

Reviewer 3: ```suggestion
Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. This will lead to more less bugs and security issues. It will also make it easier to implement new custom Providers.

We have also replaced the deprecated [request](https://github.com/request/request) library with [got](https://github.com/sindresorhus/got). This also removed the need for [purest](https://github.com/simov/purest).
```

---
title: "Uppy 3.0: Future-proof, conveniently easy, stable as ever" 
date: 2022-09-01
author: 
  - aduh95
  - arturi
  - mifi
  - murderlon

> image: "https://uppy.io/images/blog/3.0/uppy-3-0.jpg"

Reviewer 4: Any reason to use a JPEG image instead of a PNG one?

Author: Usually size, in this case PNG was larger, so why not.

published: true
---

> ![](/images/blog/3.0/uppy-3-0.jpg)

Reviewer 4: ```suggestion
![Screenshot of Uppy 3.0.0 UI](/images/blog/3.0/uppy-3-0.png)
```

## Less integration code with the new remote sources preset

We’ve introduced a new Uppy preset plugin, [`@uppy/remote-sources`](/docs/remote-sources), which combines Instagram, Facebook, Google Drive, Box, Unsplash, Dropbox, One Drive, Zoom and any other remote cloud sources that Uppy will support in the future. The plugin only works with Dashboard and allows you to enable all the above sources in a single line.

Before, you had to manually include every cloud provider / remote source:

```js
uppy
  .use(Instagram, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(GoogleDrive, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(OneDrive, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(Instagram, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(Facebook, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  .use(Unsplash, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  // ...
```

After:

```js

> uppy.use(RemoteSources, { companionUrl: 'https://companion.uppy.io' })

Reviewer 4: ```suggestion
uppy.use(RemoteSources, { companionUrl: COMPANION_URL })
```

```

> To get started, check out the [docs](/docs/remote-sources).

Reviewer 4: ```suggestion
To get started, check out the [`RemoteSources` docs](/docs/remote-sources).
```

## Native mobile camera

As you are probably aware, Uppy’s Webcam plugin provides a nice UI for desktop devices to take pictures and record videos. Up until 3.0, the same UI was utilized on mobile as well.

Since most mobile devices have a system UI for taking pictures and recording videos – which is usually better than any custom solution we can offer – we’ve introduced a new option called `mobileNativeCamera` for the Webcam plugin:

```js
uppy.use(Webcam, {
  mobileNativeCamera: isMobile({ tablet: true }),
})
```

By default, it uses the [`is-mobile`](https://github.com/juliangruber/is-mobile) package to detect mobile devices, like smartphones and tablets, but you can manually set this option to `true` or `false` if you wish. When enabled, instead of the usual Camera/Webcam Uppy UI, users will be presented with buttons that open their device's native camera interface:

> <img alt="Uppy native camera UI" src="/images/blog/3.0/native-camera.jpg">

Reviewer 4: ```suggestion
![Uppy native camera UI](/images/blog/3.0/native-camera.jpg)
```

[`videoConstraints.facingMode`](/docs/webcam/#videoConstraints) is also supported by this option via the [`capture`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/capture) attribute (in the browsers and devices that respect it).

```js
uppy.use(Webcam, {
  videoConstraints: {
    facingMode: 'user',
  },
})
```

Same applies to [`modes`](/docs/webcam/#modes), so you can enable only video or only photos.

## Devotedly stable

Since the Uppy 2.7.0 release, **we’ve fixed over 95 bugs!** This includes bugs that made uploading less stable, documentation corrections, and dependency upgrades.

## Companion turns 4.0

Everyone’s favorite Companion also received some love.

### Streaming upload

> Streaming upload can now also be enabled in Companion when using Tus. This comes with greatly improved upload speeds and allows uploading potentially hundreds of gigabytes without needing a large server storage. This feature was also available before Companion 4, but it didn’t work with Tus yet.

Reviewer 3: ```suggestion
Streaming upload can now also be enabled in Companion when using Tus. This comes with greatly improved upload speeds and allows uploading potentially hundreds of gigabytes without needing a large server storage. [We found that](https://github.com/transloadit/uppy/pull/4046#issuecomment-1235697937) this improves speeds by about 37% for a Google Drive upload of a 1GB file. This feature was also available before Companion 4, but it didn’t work with Tus until now.
```

Reviewer 4: -1 on using `We found that` as the text for the link, I'd much rather prefer to have `(source)` at the end of the sentence.

With streaming upload disabled (default), the whole file will be downloaded first. The upload will then start when the download has completely finished.

When streaming upload is enabled, Companion will start downloading the file from the provider (e.g., Google Drive), while at the same time starting the upload to the destination (e.g., Tus), and sending every chunk of data consecutively.

> For more information, [see the docs](/docs/companion/).

Reviewer 4: ```suggestion
For more information, see the [Companion docs](/docs/companion/).
```

### Event emitter

Companion now has a server-side event emitter that can be used to detect when uploads start, finish and fail, without having to depend on the client (so it works even when users have closed their browser).

Example code:

```js
const { app, emitter } = companion.app(options)

emitter.on('upload-start', ({ token }) => {
  console.log('Upload started', token)

  function onUploadEvent ({ action, payload }) {
    if (action === 'success') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.log('Upload finished', token, payload.url)
    } else if (action === 'error') {
      emitter.off(token, onUploadEvent) // avoid listener leak
      console.error('Upload failed', payload)
    }
  }
  emitter.on(token, onUploadEvent)
})
```

### Internal improvements

Companion’s internal request code and Providers have been rewritten to `async`/`await`, which has greatly simplified and reduced the amount of code. This will lead to less bugs and security issues, and make it easier to implement new custom Providers.

> We have also replaced the deprecated [request](https://github.com/request/request) library with [got](https://github.com/sindresorhus/got). This also removed the need for [purest](https://github.com/simov/purest).

Reviewer 4: ```suggestion
We have also replaced the deprecated [`request`](https://github.com/request/request) library with [`got`](https://github.com/sindresorhus/got). This also removed the need for [`purest`](https://github.com/simov/purest).
```

### And more!

* [Compressor](/docs/compressor) is officially stable. The Compressor plugin for Uppy optimizes images (JPEG, PNG), saving on average up to 60% in size (roughly 18 MB for 10 images). It has proven itself through extensive battle testing.
* [AWS S3 Multipart](/docs/aws-s3-multipart) also became more stable this release, as a result of us hammering out some of the last edge cases with lots of files and/or huge files.
* React 18 is now supported in [`@uppy/react`](/docs/react), Vue 3 is supported for [`@uppy/vue`](/docs/vue), and Angular 14 for [`@uppy/angular`](/docs/angular).

> * Our [`examples`](https://github.com/transloadit/uppy/tree/main/examples) have received a fresh round of updates to keep them relevant.

Reviewer 4: ```suggestion
* Our [examples](https://github.com/transloadit/uppy/tree/main/examples) have received a fresh round of updates to keep them relevant.
```

***

> ![](/images/blog/3.0/dog-enjoys.gif)

Reviewer 4: ```suggestion
![Animate picture of a visibly happy dog getting a head massage](/images/blog/3.0/dog-enjoys.gif)
```

> For more details, see the full [changelog](https://github.com/transloadit/uppy/blob/master/CHANGELOG.md#1300) and the [migration guide](/docs/migration-guides.html).

Reviewer 4: ```suggestion
For more details, see the full [changelog](https://github.com/transloadit/uppy/blob/HEAD/CHANGELOG.md#1300) and the [migration guide](/docs/migration-guides.html).
```

  .use(Unsplash, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
  // ...
```

After:

```js
uppy.use(RemoteSources, { companionUrl: 'https://companion.uppy.io' })
```

To get started, checkout the [docs](/docs/remote-sources).

## Robodog ends its service

Uppy is flexible and extensible through plugins. But the integration code could sometimes be daunting. This is what brought [Robodog](/docs/robodog/) to life. An alternative with the same features, but with a more ergonomic and minimal API.

But, it didn’t come without its own set of new problems:

* It tries to do the exact same, but it looks like a different product.
* It’s confusing for users whether they want to use Robodog or Uppy directly.
* Robodog is more ergonomic because it’s limited. When you hit such a limit, you need to refactor everything to Uppy with plugins.

This has now led us to deprecating Robodog and embrace Uppy for its strong suits: modularity and flexibility. At the same time, we also introduced something to take away some repetitive integration code: [`@uppy/remote-sources`](/docs/remote-sources).

> To mimic the Robodog implementation with all its features, you can use the code snippet below. But chances are Robodog did more than you need so feel free to remove things or go through the [list of plugins](/docs/plugins/) and install and use the ones you need.
> 
> > You can also checkout how we migrated the Robodog example ourselves in this [commit](https://github.com/transloadit/uppy/commit/089aaed615c77bafaf905e291b6b4e82aaeb2f6f).
> 
> ```js
> import Uppy from '@uppy/core'
> import Dashboard from '@uppy/dashboard'
> import RemoteSources from '@uppy/remote-sources'
> import Webcam from '@uppy/webcam'
> import ScreenCapture from '@uppy/screen-capture'
> import ImageEditor from '@uppy/image-editor'
> import Audio from '@uppy/audio'
> import Transloadit, { COMPANION_URL, COMPANION_ALLOWED_HOSTS } from '@uppy/transloadit'
> 
> import '@uppy/core/dist/style.css'
> import '@uppy/dashboard/dist/style.css'
> import '@uppy/webcam/dist/style.css'
> import '@uppy/audio/dist/style.css'
> import '@uppy/screen-capture/dist/style.css'
> import '@uppy/image-editor/dist/style.css'
> 
> const uppy = new Uppy()
>   .use(Dashboard, { inline: true, target: '#app' })
>   .use(RemoteSources, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
>   .use(Webcam, { target: Dashboard })
>   .use(Audio, { target: Dashboard })
>   .use(ScreenCapture, { target: Dashboard })
>   .use(ImageEditor, { target: Dashboard })
>   .use(Transloadit, {
>     async getAssemblyOptions (file) {
>       // This is where you configure your auth key, auth secret, and template ID
>       // https://uppy.io/docs/transloadit/#getAssemblyOptions-file
>       //
>       // It is important to set the secret in production:
>       // https://transloadit.com/docs/topics/signature-authentication/
>       const response = await fetch('/some-endpoint')
>       return response.json()
>     },
>   })
> ```

Reviewer 4: ```suggestion
Are you using Robodog in your code base? You can checkout out the [Migration Guide](https://uppy.io/docs/migration-guides.html#Migrate-from-Robodog-to-Uppy-plugins) to see how we can use Uppy plugins to replicate Robodog features with Uppy plugins. We have committed ourselves to extends Robodog support for one year to give our users more time to migrate.
```

---

<!--more-->

## ESM

Following the footsteps of many packages, we now only ship Uppy core and its plugins as ECMAScript Modules (ESM). On Uppy 2.x, we were shipping CommonJS.

If are already using ESM yourself, or are using the CDN builds, nothing changes for you! If you are using CommonJS, you might need to add some tooling for everything to work, or you might want to refactor your codebase to ESM – refer to the [Pure ESM package gist](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) for added information and help on how to do that.

<img class="border" alt="Uppy Compressor plugin showing compressed images notification" src="/images/blog/2.4-2.7/compressor.jpg">

## Remote Sources

We’ve introduced a new Uppy plugin/preset that combines Instagram, Facebook, Google Drive, Box, Unsplash, Dropbox, One Drive, Zoom and any future remote cloud sources that Uppy will support. It only works with Dashboard and let’s you enable all the above sources in only line:

```js
uppy.use(RemoteSoruces, { companionUrl: 'https://companion.uppy.io' }) // Remote Soruces need Companion server instance to work, you can get one hosted from Transloadit.com or self-host it
```

> ## Robodog is Deprecated

Reviewer 4: ```suggestion
## Robodog is deprecated
```

Uppy is flexible and extensible through plugins. But the integration code could sometimes be daunting. This is what brought [Robodog](/docs/robodog/) to life. An alternative with the same features, but with a more ergonomic and minimal API.

> But, it didn’t come with its own set of new problems:

Reviewer 4: ```suggestion
But, it didn’t come without its own set of new problems:
```

This has now led us to deprecating Robodog and embrace Uppy for its strong suits: modularity and flexibility. At the same time, we also introduced something to take away some repetitive integration code: [`@uppy/remote-sources`](/docs/remote-sources).

> Are you using Robodog in your code base? You can checkout out the [migration guide](https://uppy.io/docs/migration-guides.html#Migrate-from-Robodog-to-Uppy-plugins) to see how you can transition to Uppy plugins without loosing functionality. We have committed ourselves to extend Robodog support for one year to give our users more time to migrate.

Reviewer: ```suggestion
Are you using Robodog in your code base? You can check out the [migration guide](https://uppy.io/docs/migration-guides.html#Migrate-from-Robodog-to-Uppy-plugins) to see how you can transition to Uppy plugins without losing functionality. We have committed ourselves to extend Robodog support for one year to give our users more time to migrate.
```
